update to support axios v0.18.0 and solve race condition

gabrielpoca opened the pull request:
The most important change in this commit is the digest function. In version 0.13.0 there was a
[commit in axios](https://github.com/axios/axios/commit/e833a2f7e4a5ec32516e3f17763374c0377a4323) moving the request transformers to after the request interceptors. This means that the request interceptors will see a different configuration from the response interceptor, so the output of the digest function will be different in these interceptors, breaking axios-vcr functionality. To solve this we can simply apply the same request transformers to the configuration in the request interceptor.

gabrielpoca committed:
update to support axios v0.18.0

The most important change in this commit is the digest function. In
version 0.13.0 there was a
[commit in axios](https://github.com/axios/axios/commit/e833a2f7e4a5ec32516e3f17763374c0377a4323)
moving the request transformers to after the request interceptors. This
means that the request interceptor will see a different confguration
from the response interceptor, so the output of the digest function will
be different in them, breaking axios-vcr functionality.
To solve this we can simply apply the same request transformers to the
configuration in the request interceptor.

nettofarah commented:
Hey, @gabrielpoca!
Thanks for the submission

Wouldn’t this patch override requests with the same params but different headers?

What happens if I make two requests that respond with different response headers?

gabrielpoca commented:
@nettofarah yes, you're right. A coworker was telling me that other libraries do the same thing. For me that's an acceptable compromise, but I don't know if this is something you want. I can add it to the README if you think it's enough.

gabrielpoca commented:
@nettofarah this actually doesn't solve everything. The data is also transformed in the dispatchRequest middleware, so post requests with a body will have different digests in the request and response interceptors. I was thinking that maybe I can apply the same transformer functions in the request interceptor to create a config similar to the one created by the dispatchRequest.

gabrielpoca committed:
allow concurrent requests

there's a race condition updating the jsondb file when two requests are
made simulatenously, where one request overrides the output of the
other.

gabrielpoca commented:
@nettofarah I updated the code and added another commit to solve a race condition I was having. I don't like the solution because it's too coupled to axios's internals, but I don't have a better idea.

gabrielpoca commented:
Hey @nettofarah I don't wan't to bother you, but I would like to know if you're interested in merging this pull request.

nettofarah commented:
@gabrielpoca I just merged a PR that upgrades the library to work with 0.19.
Would you like write access to the repo so you can get some of the other changes in?

gabrielpoca commented:
@nettofarah thank you! Yes! This project is really useful to me and I have "hacked" it all over the place to make it work in different scenarios xd but I would still like to have your review on stuff as I'm afraid of breaking functionality for other users.

nettofarah commented:
Dope! you're a maintainer now.
I'll go ahead and close this PR, but feel free to open a new one if you'd like.

Obrigado! :)
